File: Data_Structure/DS_Using_Python/Extended_Binary_Tree.py
""" Construct below tree
             1
           /   \
          /     \
         2       3
        / \     /  \
       4   5   6  -1
     /  \  /\ / \
   -1 -1-1 -1  -1
    """
class ExtendedBinaryTree:
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None
		
	#Check if the tree is complete
	def traverse(self, tree):
		if tree is not None:
			self.traverse(tree.left)
			print(tree.data,end=' ')
			self.traverse(tree.right)
		else:
			tree = ExtendedBinaryTree(-1)
			print(tree.data,end=' ')
		
	#insert data in left child node
	def insertLeft(self, newNode):
		if self.left is None:
			self.left = ExtendedBinaryTree(newNode)
		else:
			tree = ExtendedBinaryTree(newNode)
			tree.left = self.left
			self.left = tree
		
	#insert data in right child node
	def insertRight(self, newNode):
		if self.right is None:
			self.right = ExtendedBinaryTree(newNode)
		else:
			tree = ExtendedBinaryTree(newNode)
			tree.right = self.right
			self.right = tree
		
root = ExtendedBinaryTree(1)
root.insertLeft(2)
root.insertRight(3)
root.insertLeft(4)
#root.insertRight(5)
#root.insertLeft(6)

root.traverse(root)

File: Data_Structure/Trees/Extended_Binary_Tree.c
#include<stdio.h>
#define bool int

struct node
{
    int data;
    struct node* left;
    struct node* right;
};

//Creating a new node
struct node* createNode(int item)
{
    struct node* newNode = (struct node*)malloc(sizeof(struct node));
    newNode -> data = item;
    newNode -> left = NULL;
    newNode -> right = NULL;
    return newNode;
}

//Function for inorder traversal
void traverse(struct node* root)
{
    if(root != NULL)
    {
        traverse (root -> left);
        printf("%d ", root -> data);
        traverse (root -> right);
    }
    else
    {
        //Making external nodes 
        root = createNode(-1);
        printf("%d ", root -> data);
    }
}

//Insert data in left child node
struct node* insertLeft(struct node* root, int item)
{
    root -> left = createNode(item);
    return root -> left;
}


//Insert data in right child node
struct node* insertRight(struct node* root, int item)
{
    root -> right = createNode(item);
    return root -> right;
}

int main()
{
    struct node* root = createNode(2);
    insertLeft(root, 5);
    insertRight(root, 7);
    insertLeft(root -> left, 8);
    insertRight(root -> left, 82);
    insertLeft(root -> right, 10);
    
    traverse(root);
    return 0;
}